vopr: false positive during upgrades

matklad opened the pull request:
Normally, executing upgrade is a noreturn function. In the simulator, it
is not, and instead the replica is restarted on the next tick.

The problem with this is that there might be some callbacks scheduled on
the storage, which fire before the actual upgrade happens, and observe
an inconsistent state (as, aftre transitioning to a new release replica
doesn't bother to repair local invariants).

Cancel the storage immediately upon scheduling upgrades to prevent those
callbacks from firing.

Seed: 2843879175163466061

sentientwaffle commented on src/vsr/replica.zig:
```diff
@@ -7764,7 +7764,7 @@ pub fn ReplicaType(
         }
 
-        fn transition_to_recovering_head(self: *Self) void {
+        pub fn transition_to_recovering_head(self: *Self) void {
```
Was this included by mistake?

sentientwaffle commented:
Another possibility is to reset the replica's storage during `replica_release_execute_soon`.

```
            cluster.storages[replica.replica].reset();
```

(It will be reset when we crash the replica on the next tick anyways. The reason we don't just crash the replica on the same tick as the release transition is that it would require extra branching when we are transitioning during `Replica.open`, so this was simpler.)

matklad committed:
vopr: upgrades cancel async work properly

Normally, executing upgrade is a noreturn function. In the simulator, it
is not, and instead the replica is restarted on the next tick.

The problem with this is that there might be some callbacks scheduled on
the storage, which fire before the actual upgrade happens, and observe
an inconsistent state (as, aftre transitioning to a new release replica
doesn't bother to repair local invariants).

Cancel the storage immediately upon scheduling upgrades to prevent those
callbacks from firing.

Seed: 2843879175163466061

matklad commented:
>Another possibility

Yup, that seems much better, thanks!

matklad commented:
Let's get that in to clear popular VOPR failures.
